### PR TITLE
Services

### PR DESCRIPTION
**Pull request by willmcgugan**
### What this PR solves
- Adds 2 new services; a file transfer service, and a run command service.

### How it is done
- The two services are similar in that they both open a file object, either a physical file, or a stdout handle from a process. Then they write it to an m2m channel until the file ends or the port is closed.

### What to look out for
- The subprocess module docs are as clear as mud, so check if I have missed anything.

### Review
- [X] Ready for review

### CODE REVIEW

File: dataplicity/m2m/commandservice.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                command,
+                stdout=subprocess.PIPE,
+                shell=True
+            )
+        except OSError as error:
+            log.warning('%r command failed; %s', self, error)
+            return
+
+        bytes_sent = 0
+        fh = process.stdout.fileno()
```
Comment (willmcgugan) [inline dataplicity/m2m/commandservice.py]: This loop uses select and low level reads in order to poll for the channel closing, which would be the remote ends signal to end the process.

File: dataplicity/m2m/fileservice.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        except Exception:
+            log.exception("error running %r", self)
+
+    def _run_service(self, channel, path):
+        """Send a file over a port."""
+        log.debug("%r started", self)
+        bytes_sent = 0
+        try:
+            with open(path, 'rb') as read_file:
+                read = partial(read_file.read, self.CHUNK_SIZE)
```
Comment (willmcgugan) [inline dataplicity/m2m/fileservice.py]: This is a nice pattern for reading files, using the little known second parameter of `iter`.

File: dataplicity/m2mmanager.py [hunk 3]
```diff
@@ -153,8 +155,17 @@ def on_instruction(self, sender, data):
             m2m_port = data['m2m_port']
             self.client.port_forward.redirect_port(device_port, m2m_port)
         elif action == 'reboot-device':
-            log.debug('reboot requested')
             self.reboot()
+        elif action == 'read-file':
```
Comment (willmcgugan) [inline dataplicity/m2mmanager.py]: The number of actions is getting a little unwieldily, and more are planned. Will consider some kind of dispatch mechanism in a future update.

**Comment by willmcgugan** [comment]
@wildfoundry/web-developers for your attn

**Comment by barfle** [comment]
Modified to make codacy optional.  Is it another false-pos?

File: dataplicity/m2mmanager.py (hunk 3 above)
Comment (barfle) [inline dataplicity/m2mmanager.py]: 👍

File: dataplicity/m2m/commandservice.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
+"""
+
+An M2M Service to run commands.
+
+The stdout it sent a line at a time, until the command is finished,
```
Comment (meshy) [inline dataplicity/m2m/commandservice.py]: ~it~~ _is_ sent

**Comment by willmcgugan** [comment]
@barfle codacy is not wrong, but I'm aware of the risks, and there isn't much alternative.

**Comment by barfle** [comment]
OK, fine by me. No objection

**Comment by willmcgugan** [comment]
@wildfoundry/web-developers anyone else want a piece of the action? @toudi ?

File: dataplicity/m2m/fileservice.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        return self._repr
+
+    def run_service(self, channel, path):
+        """Run the thread and log exceptions."""
+        try:
+            self._run_service(channel, path)
+        except Exception:
+            log.exception("error running %r", self)
+
+    def _run_service(self, channel, path):
```
Comment (toudi) [inline dataplicity/m2m/fileservice.py]: I'd say it is a little misleading that here the overriden _run_service accepts the file path as the second argument whereas in the original command it awaits the command as the second one. Not a deal breaker though

**Comment by toudi** [comment]
I like what I see. I'd say that the iter(read, b'') would be nicer with a little explanation (there is actually a hidden for-loop there) but on the other hand, one simply has to read carefuly.

As always, pure beauty pouring out of the screen, @willmcgugan :+1:

File: dataplicity/m2m/fileservice.py (hunk 5 above)
Comment (willmcgugan) [inline dataplicity/m2m/fileservice.py]: Sorry, don't follow. Can you elaborate?

Comment (toudi) [inline dataplicity/m2m/fileservice.py]: Right... on the second look, I think it's my fault entirely. Initially I thought that the CommandService was the base class for the rest of the commands and so when I saw two functions with identical names, I assumed that they were overrides. I can now see that these are completely separate services. Sorry for confusion :(

**Comment by willmcgugan** [comment]
Cool. Going to release this agent.

**Pull request merged**